### ISSUE TITLE
WebGL renderer does not support tile-drawn / tile-drawing

### ISSUE THREAD

**Issue by thec0keman** [openseadragon/openseadragon#2483]
It appears from the source code that the WebGL drawer does not support the tile-drawn / tile-drawing events:
https://github.com/openseadragon/openseadragon/blob/master/src/webgldrawer.js#L102

The tile-ready event seems to be the equivalent.

Two things:
1) It would be helpful to update the documentation for these events to indicate this distinction.  (Given the scope of changes with the WebGL renderer it might even be helpful to have a section dedicated to upgrading to it?)

2) Should rejectEventHandler be a bit noisier about rejected callbacks?   It looks like from the code it should be logging something, but we aren't seeing any errors logged for this.

**Comment by pearcetm** [comment]
The reason that these events aren't supported in the new drawer is that the webgl drawing pipeline doesn't allow programmatically interacting with the pixel data during drawing like the canvas drawer does (which is what those events allow). Exactly what the preferred way to do this will be will depend on the in-progress cache overhaul PR (https://github.com/openseadragon/openseadragon/pull/2407). Until this is figured out, I don't think we know what a new section of the documentation site should look like. Your input would be very welcome though!

The JSDocs have already been updated as part of the webgl PR to include the info about `tile-drawn` and `tile-drawing` not being supported by the webgl drawer. I would guess the online docs probably won't reflect this until the next release, however... is that right, @iangilman?

In addition to these changes to the docs, there's also a new mechanism that logs a warning if you try to listen for one of these non-supported events - see https://github.com/openseadragon/openseadragon/pull/2310#issuecomment-1912423632. The hope is that this will help developers notice this change ASAP if they happen to be using `tile-drawing` or `tile-drawn`.

**Comment by thec0keman** [comment]
Thanks for the explanation @pearcetm !  Is there any special setup required to see the exceptions when listening to a non-supported event?  I saw that code and was just surprised I hadn't seen any warnings while debugging tile-drawn callbacks not firing.

**Comment by pearcetm** [comment]
> Is there any special setup required to see the exceptions when listening to a non-supported event?

There shouldn't be... in the drawer comparison demo (at http://localhost:8000/test/demo/drawercomparison.html, if you've run `grunt dev`) I get the following when adding handlers to the canvas drawer (`viewer`) and the webgl drawer (`viewer2`) in the console:
<img width="718" alt="image" src="https://github.com/openseadragon/openseadragon/assets/12499997/bd6586cc-ec68-4b69-a285-56fa08742f29">

How are you adding handlers for this event? What are you seeing in the console?

**Comment by iangilman** [comment]
> I would guess the online docs probably won't reflect this until the next release, however... is that right, @iangilman?

Correct!

**Comment by Aiosa** [comment]
We will have to documment a lot of stuff. My current idea is that instead of `tile-drawn` and related canvas drawer oriented events, a generic 'invalidate tile' pipeline will be added. This will allow you to 
 - say 'tiles are outdated' https://github.com/openseadragon/openseadragon/blob/47419a090acb657a0c9bb9047a2b5493c5a309e7/test/demo/filtering-plugin/plugin.js#L119
 - process data in reaction to the above https://github.com/openseadragon/openseadragon/blob/47419a090acb657a0c9bb9047a2b5493c5a309e7/test/demo/filtering-plugin/plugin.js#L58
 - let the system automatically deal with your output data in whatever format the conversion pipeline supports

So you don't react to what a drawer is doing, but rather request an update event and do your magic. The former approch does not work since
 - generic drawer can have many ways of rendering, not all rendering pipelines let you add aditional steps at will (WebGL, WebGPU...)
 - the new async cache (as it is being designed) will NOT make renderer wait so even if the event happens, your handlers will have to use the cache system (no more `context2D` prop hacks) and will have to work with async calls, so the logics will happen after the render animation anyway, not during

I was thinking whether we really need the cache to be async, we could just drop support going to `image` type (e.g. we could create canvas from image but not vice versa), which is I think the only case that needs async execution. But such design (although much easier to program and 'friendlier' to the old OSD ways) is also very limiting - any type conversion that requires callback or promises would not be supported. Optionally, we could just allow users to switch between async and sync pipelines, but that would IMHO lead to a lot of confusion later on.

**Comment by iangilman** [comment]
@Aiosa Agreed! Thank you for sharing this context.

**Comment by hem4nn** [comment]
hey @iangilman , could you please assign this issue to me so that i can work on it.

**Comment by iangilman** [comment]
@hem4nn We're talking about a lot of documentation here, and I think @pearcetm is in the best position to understand what needs to be written. That said, perhaps you could help... Do you have a good sense of what we're talking about and what info needs to be documented?

**Comment by hem4nn** [comment]
yeah thats a lot, anyway i'll try to align them up and make up my first contribution, If i got any doubts sure i'll let you know about it @iangilman.

**Comment by pearcetm** [comment]
Much of the current issue is that the online docs don't yet reflect what has already been documented in the source code. The online docs will change to reflect this at the time of the next release.

The majority of what still needs to be documented will depend on https://github.com/openseadragon/openseadragon/pull/2407, since that will determine what the new APIs are that can serve as a replacement for the use-cases where `tile-drawing`/`tile-drawn` are currently used.

So, I don't think there's actually much to do right at the moment.

**Comment by oblak1** [comment]
I'm testing out code, forked from [this](https://github.com/usnistgov/OpenSeadragonFiltering/blob/master/openseadragon-filtering.js) and due to `tile-drawing` not being supported the whole code kind of doesn't work anymore. Reading the last comment @pearcetm there really isn't much we can do for now?

**Comment by iangilman** [comment]
@oblak1 In fact the cache improvement PR is now at a point where it's ready to be tested, if you're up for it!

**Comment by oblak1** [comment]
great! gonna check it out
